### DISCUSSION TITLE
postprocess for a vector of element integrals

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#18888] [Q&A General] [answered]
Hi, 

Can I integrate a function on each element and return a vector of the results where the size of the vector is equal to the number of elements in the domain? 
I know `ElementIntegralVariablePostprocessor` sums all integration results in the domain but I need the integration of each individual element to be output in a vector.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Why do you need this as a vector rather than a field (auxvariable) ? Is this for output to CSV?

Do you need this everywhere or just in a few locations?

Guillaume

**Reply by aaelmeli** [reply]
I need it everywhere (i.e for each element). 
I am working on a material inversion problem where I need to compute the gradient of an objective function w.r.t material parameter. For a general material distribution case, **each element will have distinct parameter**, for example, we would have randomly distributed (in space) Young's modulus. In this case, the size of the gradient vector = number of elements (which is equal to the number of material parameters) with  `gradient[ i ] =element_integral(adjoint_solution * material_derivative * forward_solution)` where material_derivative is zero everywhere except at that ith element.

**Reply by GiudGiud** [reply]
so will you be processing that data outside of moose? we can look at gradients on variables within moose with a custom auxkernel for example. 
I m asking because if not working outside of moose, I would use an auxvariable rather than postprocessors

**Reply by aaelmeli** [reply]
Yes, I will use it outside moose.

**Reply by GiudGiud** [reply]
This vectorpostprocessor is doing what you want, if you apply to a variable that contains the ElementWise-integral. 
https://mooseframework.inl.gov/source/vectorpostprocessors/ElementValueSampler.html

For the variable, use a CONSTANT MONOMIAL variable and use this auxkernel to populate it (with a L1 norm):
https://mooseframework.inl.gov/source/auxkernels/ElementLpNormAux.html

If the documentation for this latter one is not good enough, please let me know, I just added it last week and would welcome the feedback

**Reply by aaelmeli** [reply]
Thanks @GiudGiud 
One issue that makes this approach is not suitable is that it does not take the **sign** into consideration. So it only returns the absolute value of the integration. 

I do think the documentation is pretty well!

**Reply by GiudGiud** [reply]
Ah good point. You could modify the source code very easily to remove the abs() and create a new class in your application / your MOOSE version

**Comment by xueyang94** [comment]
The gradient of an objective function w.r.t. material can be computed using the DerivativeParsedMaterial in the Materials block, with the additional_derivative_symbols parameter. Paraview can give you the cell (elemental) data in the whole domain give the moose .e file.

**Reply by aaelmeli** [reply]
> The gradient of an objective function w.r.t. material can be computed using the DerivativeParsedMaterial in the Materials block, with the additional_derivative_symbols parameter. Paraview can give you the cell (elemental) data in the whole domain give the moose .e file.

Thanks @xueyang94 , I am learning that on moose website. 
If you have input.i file example other than that on moose website would be much appreciated.
